#!/bin/bash

# align RNA seq fastq files using STAR
# 24 CPUs, 96Gb memory required per human/mouse sample
# can map paired and single reads in a single job
# Should scale to the number of reads in FASTQ file

#PBS -P er01
#PBS -N star_align_trimmed
#PBS -l walltime=05:00:00,ncpus=432,mem=1728GB,wd
#PBS -q normal
#PBS -W umask=022
#PBS -l storage=scratch/er01
#PBS -o ./Logs/star_align_trimmed.o
#PBS -e ./Logs/star_align_trimmed.e

module load openmpi/4.0.2
module load nci-parallel/1.0.0
module load star/2.7.3a

set -e

# M = Number of tasks per node
# NCPUs = CPUs per task
M=4
NCPUS=12

INPUTS=./Inputs
paired=${INPUTS}/star_align_paired.inputs
single=${INPUTS}/star_align_single.inputs

echo "$(date): STAR alignment."

if [ -s ${paired} ]
then
	echo "$(date): Found inputs file for paired data. Performing alignment with STAR for paired data."
	SCRIPT=./star_align_paired_with_unmapped.sh
	sed "s|^|${SCRIPT} |" ${paired} > ${PBS_JOBFS}/input-file-paired
	mpirun --np $((M * PBS_NCPUS / 48)) \
		--map-by node:PE=${NCPUS} \
		nci-parallel \
		--verbose \
		--input-file ${PBS_JOBFS}/input-file-paired
fi

if [ -s ${single} ]
then
	echo "$(date): Found inputs file for single data. Performing alignment with STAR for single read data."
	SCRIPT=./star_align_single_with_unmapped.sh
	sed "s|^|${SCRIPT} |" ${single} > ${PBS_JOBFS}/input-file-single
        mpirun --np $((M * PBS_NCPUS / 48)) \
                --map-by node:PE=${NCPUS} \
                nci-parallel \
                --verbose \
                --input-file ${PBS_JOBFS}/input-file-single
fi
